Src/FluentAssertions.Net40/Xml/XmlReaderValidator.cs

```diff
@@ -0,0 +1,239 @@
+        {
+            var validationResult = Validate();
+
+            if(expectedEquivalence && validationResult != null)
+            {
+                assertion.FailWith(validationResult.FormatString, validationResult.FormatParams);
+            }
+            if(!expectedEquivalence && validationResult == null)
+            {
+                assertion.FailWith("Expected Xml to not be equivalent{reason}, but it is.");
+            }
+        }
+
+        private ValidationResult Validate()
+        {
+            subjectReader.MoveToContent();
+            expectedReader.MoveToContent();
+            while (!subjectReader.EOF && !expectedReader.EOF)
+            {
+                if (subjectReader.NodeType != expectedReader.NodeType)
+                {
+                    return new ValidationResult("Expected node of type {0} at {1}{reason}, but found {2}.",
+                        expectedReader.NodeType, CurrentLocation, subjectReader.NodeType);
+                }
+
+                ValidationResult validationResult = null;
+
+                switch (subjectReader.NodeType)
+                {
+                    case XmlNodeType.Element:
+                        validationResult = ValidateStartElement();
+                        if(validationResult != null)
+                        {
+                            return validationResult;
+                        }
+                        locationStack.Push(subjectReader.LocalName);
+                        validationResult = ValidateAttributes();
+                        break;
+                    case XmlNodeType.EndElement:
+                        // No need to verify end element, if it doesn't match
+                        // the start element it isn't valid XML, so the parser
+                        // would handle that.
+                        locationStack.Pop();
+                        break;
+                    case XmlNodeType.Text:
+                        validationResult = ValidateText();
+                        break;
+                    default:
+                        throw new NotImplementedException();
+                }
+
+                if(validationResult != null)
+                {
+                    return validationResult;
+                }
+
+                subjectReader.Read();
+                expectedReader.Read();
+
+                subjectReader.MoveToContent();
+                expectedReader.MoveToContent();
+            }
+
+            if (!expectedReader.EOF)
+            {
+                return new ValidationResult("Expected {0}{reason}, but found end of document.",
+                    expectedReader.LocalName);
+            }
+
+            if (!subjectReader.EOF)
+            {
+                return new ValidationResult("Expected end of document{reason}, but found {0}.",
+                    subjectReader.LocalName);
+            }
+
+            return null;
+        }
+
+
+        class AttributeData
```

Comment:
This class is only used internally in `XmlReaderValidator` and never visible to the outside. I put it here to make it private to the class. Do you prefer it in an separate file as `internal`?